File: coding_patterns/Pattern 01 - Sliding Window/m - Fruits into Baskets.py
def totalFruit(fruits: list[int]) -> int:
    start: int  = 0
    max_length = 0
    map: dict[int, int] = {}
    
    for i in range(len(fruits)):
        if fruits[i] not in map:
            map[fruits[i]] = 0
        map[fruits[i]] += 1
        
        if len(map) > 2:
            map[fruits[start]] -= 1
            
            if map[fruits[start]] == 0:
                del map[fruits[start]]
            start += 1
        max_length = max(max_length, i - start + 1)
    return max_length
if __name__ == '__main__':
    print(totalFruit([3,3,3,1,2,1,1,2,3,3,4]))

File: coding_patterns/Pattern 01 - Sliding Window/m - Fruits into Baskets.ts
function totalFruit(fruits: number[]): number {
    let start: number = 0;
    let max_length: number = 0;
    let map: Record<number, number> = {};

    for (let i: number = 0; i < fruits.length; i++) {
        if (Object.keys(map).length < 2) {
            map[fruits[i]] = 0;
        }
        map[fruits[i]]++;

        while (Object.keys(map).length > 2) {
            map[fruits[start]]--;
            if (map[fruits[start]] === 0) {
                delete map[fruits[start]];
            }
            start++;
        }
        max_length = Math.max(max_length, i - start + 1);
    }
    return max_length;

};

console.log(totalFruit([1, 2, 1]));